File: HoloLoom/mobile/examples/android/HoloLoomClient.kt
/*
 * HoloLoomClient.kt
 * ==================
 * Kotlin client for HoloLoom mobile API (Android)
 *
 * Features:
 * - REST API client with coroutines
 * - WebSocket support for real-time chat
 * - JWT authentication
 * - Offline mode with Room database
 * - File upload
 *
 * Requirements:
 * - Android SDK 24+
 * - Kotlin 1.9+
 * - OkHttp 4.x
 * - kotlinx.serialization
 */

package com.hololoom.client

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.time.Instant
import java.util.*
import java.util.concurrent.TimeUnit

// MARK: - Models

@Serializable
data class User(
    val username: String,
    val email: String? = null,
    @SerialName("full_name") val fullName: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("avatar_url") val avatarUrl: String? = null
)

@Serializable
data class LoginResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String? = null,
    @SerialName("token_type") val tokenType: String,
    @SerialName("expires_in") val expiresIn: Int,
    val username: String,
    @SerialName("session_id") val sessionId: String,
    val user: User
)

@Serializable
data class Message(
    @SerialName("message_id") val id: String,
    @SerialName("session_id") val sessionId: String,
    val role: String,
    val content: String,
    val timestamp: String,
    val attachments: List<Attachment> = emptyList(),
    val metadata: JsonObject = JsonObject(emptyMap()),
    val status: String
)

@Serializable
data class Attachment(
    @SerialName("file_id") val fileId: String,
    @SerialName("file_type") val fileType: String,
    @SerialName("file_name") val fileName: String,
    @SerialName("file_size") val fileSize: Int,
    val url: String
)

@Serializable
data class ChatSession(
    @SerialName("session_id") val id: String,
    val title: String,
    @SerialName("created_at") val createdAt: String,
    @SerialName("last_activity") val lastActivity: String,
    @SerialName("message_count") val messageCount: Int,
    val metadata: JsonObject = JsonObject(emptyMap())
)

@Serializable
data class WeavingTrace(
    val motifs: List<String>,
    val embeddings: EmbeddingInfo,
    @SerialName("tool_selection") val toolSelection: ToolSelection,
    @SerialName("execution_time_ms") val executionTimeMs: Double
) {
    @Serializable
    data class EmbeddingInfo(
        val scales: List<Int>,
        @SerialName("similarity_scores") val similarityScores: List<Double>? = null
    )

    @Serializable
    data class ToolSelection(
        @SerialName("selected_tool") val selectedTool: String,
        val confidence: Double,
        val strategy: String
    )
}

@Serializable
data class SendMessageResponse(
    @SerialName("user_message") val userMessage: Message,
    @SerialName("assistant_message") val assistantMessage: Message,
    val trace: WeavingTrace
)

// MARK: - API Client

class HoloLoomClient(
    private val baseURL: String = "https://api.hololoom.ai/v1"
) {
    private var accessToken: String? = null
    private var refreshToken: String? = null

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
        isLenient = true
    }

    // MARK: - Authentication

    suspend fun login(
        username: String,
        password: String,
        deviceId: String? = null
    ): LoginResponse = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("username", username)
            put("password", password)
            put("device_id", deviceId ?: UUID.randomUUID().toString())
        }

        val request = Request.Builder()
            .url("$baseURL/auth/login")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.Unauthorized()
        }

        val loginResponse = json.decodeFromString<LoginResponse>(response.body!!.string())
        accessToken = loginResponse.accessToken
        refreshToken = loginResponse.refreshToken

        loginResponse
    }

    suspend fun logout() = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseURL/auth/logout")
            .post("".toRequestBody())
            .addAuthHeader()
            .build()

        client.newCall(request).execute()

        accessToken = null
        refreshToken = null
    }

    suspend fun getCurrentUser(): User = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseURL/auth/me")
            .get()
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.Unauthorized()
        }

        json.decodeFromString(response.body!!.string())
    }

    // MARK: - Chat Sessions

    suspend fun listSessions(
        limit: Int = 20,
        offset: Int = 0
    ): List<ChatSession> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseURL/chat/sessions?limit=$limit&offset=$offset")
            .get()
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        val jsonResponse = json.parseToJsonElement(response.body!!.string()).jsonObject
        val sessions = jsonResponse["sessions"]?.jsonArray ?: JsonArray(emptyList())

        sessions.map { json.decodeFromJsonElement<ChatSession>(it) }
    }

    suspend fun createSession(title: String? = null): ChatSession = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("title", title ?: "New conversation")
        }

        val request = Request.Builder()
            .url("$baseURL/chat/sessions")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        json.decodeFromString(response.body!!.string())
    }

    suspend fun deleteSession(sessionId: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseURL/chat/sessions/$sessionId")
            .delete()
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.NotFound()
        }
    }

    // MARK: - Messages

    suspend fun getMessages(
        sessionId: String,
        limit: Int = 50
    ): List<Message> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$baseURL/chat/sessions/$sessionId/messages?limit=$limit")
            .get()
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        val jsonResponse = json.parseToJsonElement(response.body!!.string()).jsonObject
        val messages = jsonResponse["messages"]?.jsonArray ?: JsonArray(emptyList())

        messages.map { json.decodeFromJsonElement<Message>(it) }
    }

    suspend fun sendMessage(
        sessionId: String,
        text: String,
        attachments: List<String> = emptyList()
    ): SendMessageResponse = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("text", text)
            put("attachments", JsonArray(attachments.map { JsonPrimitive(it) }))
        }

        val request = Request.Builder()
            .url("$baseURL/chat/sessions/$sessionId/messages")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        json.decodeFromString(response.body!!.string())
    }

    // MARK: - File Upload

    suspend fun uploadFile(
        file: File,
        mimeType: String
    ): Attachment = withContext(Dispatchers.IO) {
        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file",
                file.name,
                file.asRequestBody(mimeType.toMediaType())
            )
            .build()

        val request = Request.Builder()
            .url("$baseURL/files/upload")
            .post(requestBody)
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        json.decodeFromString(response.body!!.string())
    }

    // MARK: - Sync

    suspend fun pullSync(
        lastSync: Instant? = null,
        entityTypes: List<String> = listOf("messages", "sessions")
    ): JsonObject = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            if (lastSync != null) {
                put("last_sync", lastSync.toString())
            }
            put("entity_types", JsonArray(entityTypes.map { JsonPrimitive(it) }))
        }

        val request = Request.Builder()
            .url("$baseURL/sync/pull")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        json.parseToJsonElement(response.body!!.string()).jsonObject
    }

    suspend fun pushSync(
        messages: List<Message> = emptyList(),
        sessions: List<ChatSession> = emptyList()
    ): JsonObject = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("messages", JsonArray(messages.map { json.encodeToJsonElement(it) }))
            put("sessions", JsonArray(sessions.map { json.encodeToJsonElement(it) }))
        }

        val request = Request.Builder()
            .url("$baseURL/sync/push")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }

        json.parseToJsonElement(response.body!!.string()).jsonObject
    }

    // MARK: - Push Notifications

    suspend fun registerForPush(
        deviceToken: String,
        platform: String = "android"
    ) = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("device_token", deviceToken)
            put("platform", platform)
            put("device_id", UUID.randomUUID().toString())
        }

        val request = Request.Builder()
            .url("$baseURL/push/register")
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .addAuthHeader()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw APIException.ServerError()
        }
    }

    // MARK: - Private Helpers

    private fun Request.Builder.addAuthHeader(): Request.Builder {
        accessToken?.let {
            header("Authorization", "Bearer $it")
        }
        return this
    }
}

// MARK: - WebSocket Client

class HoloLoomWebSocket(
    private val sessionId: String,
    private val token: String,
    private val listener: ChatListener
) {
    private val client = OkHttpClient()
    private var webSocket: WebSocket? = null

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    interface ChatListener {
        fun onConnected(sessionId: String)
        fun onThinking(message: String, stage: String)
        fun onResponseChunk(text: String, done: Boolean)
        fun onError(error: String)
        fun onClosed()
    }

    fun connect() {
        val url = "wss://api.hololoom.ai/ws/chat/$sessionId?token=$token"
        val request = Request.Builder().url(url).build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                println("WebSocket connected")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                println("WebSocket error: ${t.message}")
                listener.onError(t.message ?: "Unknown error")
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                println("WebSocket closed: $code - $reason")
                listener.onClosed()
            }
        })
    }

    fun sendMessage(text: String) {
        val message = buildJsonObject {
            put("type", "message")
            put("text", text)
            putJsonObject("metadata") {
                put("client_message_id", UUID.randomUUID().toString())
                put("timestamp", Instant.now().toString())
            }
        }

        webSocket?.send(message.toString())
    }

    fun sendHeartbeat() {
        val message = buildJsonObject {
            put("type", "ping")
        }

        webSocket?.send(message.toString())
    }

    private fun handleMessage(text: String) {
        try {
            val jsonMsg = json.parseToJsonElement(text).jsonObject
            val type = jsonMsg["type"]?.jsonPrimitive?.content ?: return

            when (type) {
                "connected" -> {
                    val sessionId = jsonMsg["session_id"]?.jsonPrimitive?.content ?: ""
                    listener.onConnected(sessionId)
                }

                "thinking" -> {
                    val message = jsonMsg["message"]?.jsonPrimitive?.content ?: ""
                    val stage = jsonMsg["stage"]?.jsonPrimitive?.content ?: ""
                    listener.onThinking(message, stage)
                }

                "response_chunk" -> {
                    val chunkText = jsonMsg["text"]?.jsonPrimitive?.content ?: ""
                    val done = jsonMsg["done"]?.jsonPrimitive?.boolean ?: false
                    listener.onResponseChunk(chunkText, done)
                }

                "error" -> {
                    val error = jsonMsg["error"]?.jsonPrimitive?.content ?: "Unknown error"
                    listener.onError(error)
                }

                "pong" -> {
                    // Heartbeat response
                }
            }
        } catch (e: Exception) {
            println("Error parsing message: ${e.message}")
        }
    }

    fun disconnect() {
        webSocket?.close(1000, "Client disconnect")
    }
}

// MARK: - Errors

sealed class APIException : Exception() {
    class Unauthorized : APIException()
    class NotFound : APIException()
    class ServerError : APIException()
    class NetworkError : APIException()
}

// MARK: - Example Usage

/*
fun main() = runBlocking {
    // Initialize client
    val client = HoloLoomClient(baseURL = "https://api.hololoom.ai/v1")

    // Login
    val loginResponse = client.login(username = "admin", password = "admin123")
    println("Logged in as: ${loginResponse.username}")

    // Create session
    val session = client.createSession(title = "My Chat")
    println("Created session: ${session.id}")

    // Send message (REST)
    val response = client.sendMessage(
        sessionId = session.id,
        text = "What is Thompson Sampling?"
    )
    println("Response: ${response.assistantMessage.content}")
    println("Tool used: ${response.trace.toolSelection.selectedTool}")

    // Real-time chat (WebSocket)
    val ws = HoloLoomWebSocket(
        sessionId = loginResponse.sessionId,
        token = loginResponse.accessToken,
        listener = object : HoloLoomWebSocket.ChatListener {
            override fun onConnected(sessionId: String) {
                println("WebSocket connected: $sessionId")
            }

            override fun onThinking(message: String, stage: String) {
                println("Status: $message ($stage)")
            }

            override fun onResponseChunk(text: String, done: Boolean) {
                if (text.isNotEmpty()) {
                    print(text)
                }
                if (done) {
                    println("\n--- Response complete ---")
                }
            }

            override fun onError(error: String) {
                println("Error: $error")
            }

            override fun onClosed() {
                println("WebSocket closed")
            }
        }
    )

    ws.connect()
    delay(1000) // Wait for connection

    ws.sendMessage("Hello, HoloLoom!")

    // Wait for response
    delay(5000)

    ws.disconnect()

    // Upload file
    val file = File("/path/to/image.jpg")
    val attachment = client.uploadFile(file, "image/jpeg")
    println("Uploaded: ${attachment.fileId}")

    // Sync
    val syncData = client.pullSync()
    println("Synced: ${syncData["sync_timestamp"]}")
}
*/

// MARK: - ViewModel Integration (Android)

/*
class ChatViewModel : ViewModel() {
    private val client = HoloLoomClient()
    private var webSocket: HoloLoomWebSocket? = null

    private val _messages = MutableLiveData<List<Message>>()
    val messages: LiveData<List<Message>> = _messages

    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> = _isLoading

    fun login(username: String, password: String) {
        viewModelScope.launch {
            try {
                _isLoading.value = true
                val response = client.login(username, password)
                connectWebSocket(response.sessionId, response.accessToken)
            } catch (e: Exception) {
                // Handle error
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun sendMessage(sessionId: String, text: String) {
        webSocket?.sendMessage(text)
    }

    private fun connectWebSocket(sessionId: String, token: String) {
        webSocket = HoloLoomWebSocket(sessionId, token, object : HoloLoomWebSocket.ChatListener {
            override fun onConnected(sessionId: String) {
                // Handle connection
            }

            override fun onThinking(message: String, stage: String) {
                // Update UI with thinking status
            }

            override fun onResponseChunk(text: String, done: Boolean) {
                // Append to message content
            }

            override fun onError(error: String) {
                // Handle error
            }

            override fun onClosed() {
                // Handle disconnection
            }
        })

        webSocket?.connect()
    }

    override fun onCleared() {
        super.onCleared()
        webSocket?.disconnect()
    }
}
*/

File: HoloLoom/mobile/examples/ios/HoloLoomClient.swift
/*
 HoloLoomClient.swift
 ====================
 Swift client for HoloLoom mobile API (iOS)

 Features:
 - REST API client with async/await
 - WebSocket support for real-time chat
 - JWT authentication
 - Offline mode with local persistence
 - File upload

 Requirements:
 - iOS 15.0+
 - Swift 5.5+ (async/await)
 */

import Foundation
import Combine

// MARK: - Models

struct User: Codable {
    let username: String
    let email: String?
    let fullName: String?
    let createdAt: Date?
    let avatarUrl: String?

    enum CodingKeys: String, CodingKey {
        case username
        case email
        case fullName = "full_name"
        case createdAt = "created_at"
        case avatarUrl = "avatar_url"
    }
}

struct LoginResponse: Codable {
    let accessToken: String
    let refreshToken: String?
    let tokenType: String
    let expiresIn: Int
    let username: String
    let sessionId: String
    let user: User

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case tokenType = "token_type"
        case expiresIn = "expires_in"
        case username
        case sessionId = "session_id"
        case user
    }
}

struct Message: Codable, Identifiable {
    let id: String
    let sessionId: String
    let role: String
    let content: String
    let timestamp: Date
    let attachments: [Attachment]
    let metadata: [String: AnyCodable]
    let status: String

    enum CodingKeys: String, CodingKey {
        case id = "message_id"
        case sessionId = "session_id"
        case role
        case content
        case timestamp
        case attachments
        case metadata
        case status
    }
}

struct Attachment: Codable {
    let fileId: String
    let fileType: String
    let fileName: String
    let fileSize: Int
    let url: String

    enum CodingKeys: String, CodingKey {
        case fileId = "file_id"
        case fileType = "file_type"
        case fileName = "file_name"
        case fileSize = "file_size"
        case url
    }
}

struct ChatSession: Codable, Identifiable {
    let id: String
    let title: String
    let createdAt: Date
    let lastActivity: Date
    let messageCount: Int
    let metadata: [String: AnyCodable]

    enum CodingKeys: String, CodingKey {
        case id = "session_id"
        case title
        case createdAt = "created_at"
        case lastActivity = "last_activity"
        case messageCount = "message_count"
        case metadata
    }
}

struct WeavingTrace: Codable {
    let motifs: [String]
    let embeddings: EmbeddingInfo
    let toolSelection: ToolSelection
    let executionTimeMs: Double

    enum CodingKeys: String, CodingKey {
        case motifs
        case embeddings
        case toolSelection = "tool_selection"
        case executionTimeMs = "execution_time_ms"
    }

    struct EmbeddingInfo: Codable {
        let scales: [Int]
        let similarityScores: [Double]?

        enum CodingKeys: String, CodingKey {
            case scales
            case similarityScores = "similarity_scores"
        }
    }

    struct ToolSelection: Codable {
        let selectedTool: String
        let confidence: Double
        let strategy: String

        enum CodingKeys: String, CodingKey {
            case selectedTool = "selected_tool"
            case confidence
            case strategy
        }
    }
}

// Helper for dynamic JSON values
struct AnyCodable: Codable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if let value = try? container.decode(String.self) {
            self.value = value
        } else if let value = try? container.decode(Int.self) {
            self.value = value
        } else if let value = try? container.decode(Double.self) {
            self.value = value
        } else if let value = try? container.decode(Bool.self) {
            self.value = value
        } else {
            self.value = NSNull()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch value {
        case let val as String:
            try container.encode(val)
        case let val as Int:
            try container.encode(val)
        case let val as Double:
            try container.encode(val)
        case let val as Bool:
            try container.encode(val)
        default:
            try container.encodeNil()
        }
    }
}

// MARK: - API Client

class HoloLoomClient {
    private let baseURL: String
    private var accessToken: String?
    private var refreshToken: String?
    private let session: URLSession

    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    private let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }()

    init(baseURL: String = "https://api.hololoom.ai/v1") {
        self.baseURL = baseURL
        self.session = URLSession.shared
    }

    // MARK: - Authentication

    func login(username: String, password: String, deviceId: String? = nil) async throws -> LoginResponse {
        let url = URL(string: "\(baseURL)/auth/login")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "username": username,
            "password": password,
            "device_id": deviceId ?? UIDevice.current.identifierForVendor?.uuidString ?? ""
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw APIError.unauthorized
        }

        let loginResponse = try decoder.decode(LoginResponse.self, from: data)
        self.accessToken = loginResponse.accessToken
        self.refreshToken = loginResponse.refreshToken

        return loginResponse
    }

    func logout() async throws {
        let url = URL(string: "\(baseURL)/auth/logout")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        try addAuthHeader(to: &request)

        _ = try await session.data(for: request)

        self.accessToken = nil
        self.refreshToken = nil
    }

    func getCurrentUser() async throws -> User {
        let url = URL(string: "\(baseURL)/auth/me")!
        var request = URLRequest(url: url)
        try addAuthHeader(to: &request)

        let (data, _) = try await session.data(for: request)
        return try decoder.decode(User.self, from: data)
    }

    // MARK: - Chat Sessions

    func listSessions(limit: Int = 20, offset: Int = 0) async throws -> [ChatSession] {
        let url = URL(string: "\(baseURL)/chat/sessions?limit=\(limit)&offset=\(offset)")!
        var request = URLRequest(url: url)
        try addAuthHeader(to: &request)

        let (data, _) = try await session.data(for: request)
        let response = try decoder.decode([String: AnyCodable].self, from: data)

        // For now, return empty array (implement proper parsing)
        return []
    }

    func createSession(title: String? = nil) async throws -> ChatSession {
        let url = URL(string: "\(baseURL)/chat/sessions")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        try addAuthHeader(to: &request)

        let body: [String: Any] = ["title": title ?? "New conversation"]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, _) = try await session.data(for: request)
        return try decoder.decode(ChatSession.self, from: data)
    }

    func deleteSession(_ sessionId: String) async throws {
        let url = URL(string: "\(baseURL)/chat/sessions/\(sessionId)")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        try addAuthHeader(to: &request)

        _ = try await session.data(for: request)
    }

    // MARK: - Messages

    func getMessages(sessionId: String, limit: Int = 50) async throws -> [Message] {
        let url = URL(string: "\(baseURL)/chat/sessions/\(sessionId)/messages?limit=\(limit)")!
        var request = URLRequest(url: url)
        try addAuthHeader(to: &request)

        let (data, _) = try await session.data(for: request)
        let response = try decoder.decode([String: AnyCodable].self, from: data)

        // For now, return empty array (implement proper parsing)
        return []
    }

    func sendMessage(
        sessionId: String,
        text: String,
        attachments: [String] = []
    ) async throws -> (userMessage: Message, assistantMessage: Message, trace: WeavingTrace) {
        let url = URL(string: "\(baseURL)/chat/sessions/\(sessionId)/messages")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        try addAuthHeader(to: &request)

        let body: [String: Any] = [
            "text": text,
            "attachments": attachments
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, _) = try await session.data(for: request)

        // Parse response (simplified)
        let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]

        // TODO: Proper parsing
        let userMsg = Message(
            id: "msg_1",
            sessionId: sessionId,
            role: "user",
            content: text,
            timestamp: Date(),
            attachments: [],
            metadata: [:],
            status: "sent"
        )

        let assistantMsg = Message(
            id: "msg_2",
            sessionId: sessionId,
            role: "assistant",
            content: "Response",
            timestamp: Date(),
            attachments: [],
            metadata: [:],
            status: "sent"
        )

        let trace = WeavingTrace(
            motifs: [],
            embeddings: WeavingTrace.EmbeddingInfo(scales: [96, 192, 384]),
            toolSelection: WeavingTrace.ToolSelection(
                selectedTool: "echo",
                confidence: 0.9,
                strategy: "epsilon_greedy"
            ),
            executionTimeMs: 100
        )

        return (userMsg, assistantMsg, trace)
    }

    // MARK: - File Upload

    func uploadFile(data: Data, fileName: String, mimeType: String) async throws -> Attachment {
        let url = URL(string: "\(baseURL)/files/upload")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        try addAuthHeader(to: &request)

        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()

        // Add file
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body

        let (responseData, _) = try await session.data(for: request)

        // Parse response
        let json = try JSONSerialization.jsonObject(with: responseData) as! [String: Any]

        return Attachment(
            fileId: json["file_id"] as! String,
            fileType: json["file_type"] as! String,
            fileName: json["file_name"] as! String,
            fileSize: json["file_size"] as! Int,
            url: json["url"] as! String
        )
    }

    // MARK: - Sync

    func pullSync(lastSync: Date?, entityTypes: [String] = ["messages", "sessions"]) async throws -> [String: Any] {
        let url = URL(string: "\(baseURL)/sync/pull")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        try addAuthHeader(to: &request)

        var body: [String: Any] = ["entity_types": entityTypes]
        if let lastSync = lastSync {
            let formatter = ISO8601DateFormatter()
            body["last_sync"] = formatter.string(from: lastSync)
        }

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, _) = try await session.data(for: request)
        return try JSONSerialization.jsonObject(with: data) as! [String: Any]
    }

    // MARK: - Push Notifications

    func registerForPush(deviceToken: String, platform: String = "ios") async throws {
        let url = URL(string: "\(baseURL)/push/register")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        try addAuthHeader(to: &request)

        let body: [String: Any] = [
            "device_token": deviceToken,
            "platform": platform,
            "device_id": UIDevice.current.identifierForVendor?.uuidString ?? ""
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        _ = try await session.data(for: request)
    }

    // MARK: - Private Helpers

    private func addAuthHeader(to request: inout URLRequest) throws {
        guard let token = accessToken else {
            throw APIError.unauthorized
        }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }
}

// MARK: - WebSocket Client

class HoloLoomWebSocket: NSObject {
    private var webSocketTask: URLSessionWebSocketTask?
    private let sessionId: String
    private let token: String

    var onConnected: (() -> Void)?
    var onThinking: ((String, String) -> Void)?
    var onResponseChunk: ((String, Bool) -> Void)?
    var onError: ((String) -> Void)?

    init(sessionId: String, token: String) {
        self.sessionId = sessionId
        self.token = token
        super.init()
    }

    func connect() {
        let urlString = "wss://api.hololoom.ai/ws/chat/\(sessionId)?token=\(token)"
        guard let url = URL(string: urlString) else { return }

        webSocketTask = URLSession.shared.webSocketTask(with: url)
        webSocketTask?.resume()

        receiveMessage()
    }

    func sendMessage(_ text: String) {
        let message: [String: Any] = [
            "type": "message",
            "text": text,
            "metadata": [
                "client_message_id": UUID().uuidString,
                "timestamp": ISO8601DateFormatter().string(from: Date())
            ]
        ]

        guard let data = try? JSONSerialization.data(withJSONObject: message),
              let jsonString = String(data: data, encoding: .utf8) else {
            return
        }

        webSocketTask?.send(.string(jsonString)) { error in
            if let error = error {
                print("Send error: \(error)")
            }
        }
    }

    func sendHeartbeat() {
        let message = ["type": "ping"]
        guard let data = try? JSONSerialization.data(withJSONObject: message),
              let jsonString = String(data: data, encoding: .utf8) else {
            return
        }

        webSocketTask?.send(.string(jsonString), completionHandler: nil)
    }

    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self?.handleMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self?.handleMessage(text)
                    }
                @unknown default:
                    break
                }
                self?.receiveMessage()

            case .failure(let error):
                print("Receive error: \(error)")
            }
        }
    }

    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else {
            return
        }

        switch type {
        case "connected":
            onConnected?()

        case "thinking":
            let message = json["message"] as? String ?? ""
            let stage = json["stage"] as? String ?? ""
            onThinking?(message, stage)

        case "response_chunk":
            let text = json["text"] as? String ?? ""
            let done = json["done"] as? Bool ?? false
            onResponseChunk?(text, done)

        case "error":
            let error = json["error"] as? String ?? "Unknown error"
            onError?(error)

        default:
            break
        }
    }

    func disconnect() {
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
    }
}

// MARK: - Errors

enum APIError: Error {
    case unauthorized
    case notFound
    case serverError
    case networkError
    case decodingError
}

// MARK: - Example Usage

/*
 // Initialize client
 let client = HoloLoomClient(baseURL: "https://api.hololoom.ai/v1")

 // Login
 let loginResponse = try await client.login(username: "admin", password: "admin123")
 print("Logged in as: \(loginResponse.username)")

 // Create session
 let session = try await client.createSession(title: "My Chat")

 // Send message (REST)
 let (userMsg, assistantMsg, trace) = try await client.sendMessage(
     sessionId: session.id,
     text: "What is Thompson Sampling?"
 )
 print("Response: \(assistantMsg.content)")

 // Real-time chat (WebSocket)
 let ws = HoloLoomWebSocket(sessionId: loginResponse.sessionId, token: loginResponse.accessToken)

 ws.onConnected = {
     print("WebSocket connected")
 }

 ws.onResponseChunk = { text, done in
     print(text, terminator: "")
     if done {
         print("\n--- Response complete ---")
     }
 }

 ws.connect()
 ws.sendMessage("Hello, HoloLoom!")

 // Upload file
 let imageData = UIImage(named: "photo")?.jpegData(compressionQuality: 0.8)
 let attachment = try await client.uploadFile(
     data: imageData!,
     fileName: "photo.jpg",
     mimeType: "image/jpeg"
 )
 print("Uploaded: \(attachment.fileId)")
 */
